File: backend/app/services/mermaid_sanitizer.py
"""Mermaid syntax normalization helpers."""

import re


CODE_FENCE_RE = re.compile(r"^```(?:mermaid)?\s*|\s*```$", re.IGNORECASE)
FLOWCHART_HEADER_RE = re.compile(r"^(?:flowchart|graph)\s+", re.IGNORECASE)
EDGE_PIPE_LABEL_RE = re.compile(r"(?P<prefix>[-.=ox<>]+\|)(?P<label>[^|\n]+)(?P<suffix>\|)")
EDGE_LABEL_IN_LINE_RE = re.compile(r"(?P<connector>[-.=ox<>]{2,})\s*\|(?P<label>[^|\n]+)\|\s*")
INLINE_EDGE_TARGET_DECL_RE = re.compile(
    r"(?P<connector>\s+[-.=ox<>]{2,}\s+)"
    r"(?P<target>[A-Za-z0-9_][A-Za-z0-9_-]*(?:\[[^\n\]]+\]|\([^\n)]+\)|\{[^\n}]+\}):::[A-Za-z0-9_-]+)\s*$"
)
NODE_ID_RE = re.compile(r"^([A-Za-z0-9_][A-Za-z0-9_-]*)")
RISKY_EDGE_LABEL_RE = re.compile(r"[()\[\]{}（）;]")


def strip_mermaid_code_fences(content: str) -> str:
    """Remove optional markdown code fences around Mermaid source."""

    unfenced = CODE_FENCE_RE.sub("", content.strip().lstrip("\ufeff")).strip()
    smart_unwrapped = re.sub(r"^[“”\"']+", "", unfenced)
    smart_unwrapped = re.sub(r"[“”\"']+$", "", smart_unwrapped).strip()
    return smart_unwrapped or unfenced


def _is_flowchart(code: str) -> bool:
    for line in code.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("%%"):
            continue
        return bool(FLOWCHART_HEADER_RE.match(stripped))
    return False


def _quote_edge_label(match: re.Match[str]) -> str:
    label = match.group("label").strip()
    if len(label) >= 2 and label.startswith('"') and label.endswith('"'):
        return match.group(0)
    if not RISKY_EDGE_LABEL_RE.search(label):
        return match.group(0)
    escaped = label.replace('"', "#quot;")
    return f'{match.group("prefix")}"{escaped}"{match.group("suffix")}'


def _clean_relation_label(label: str) -> str:
    stripped = label.strip()
    if len(stripped) >= 2 and stripped.startswith('"') and stripped.endswith('"'):
        stripped = stripped[1:-1]
    return stripped.replace('"', "#quot;")


def _stabilize_flowchart_edge_labels(code: str) -> tuple[str, bool]:
    stabilized_lines: list[str] = []
    relation_index = 0
    changed = False
    declared_inline_targets: set[str] = set()

    for line in code.splitlines():
        if line.lstrip().startswith("%%"):
            stabilized_lines.append(line)
            continue
        indent_match = re.match(r"^\s*", line)
        indent = indent_match.group(0) if indent_match else ""
        body = line[len(indent) :]
        inline_target_match = INLINE_EDGE_TARGET_DECL_RE.search(body)
        if inline_target_match:
            target_decl = inline_target_match.group("target")
            target_id_match = NODE_ID_RE.match(target_decl)
            if target_id_match:
                target_id = target_id_match.group(1)
                if target_id not in declared_inline_targets:
                    stabilized_lines.append(f"{indent}{target_decl}")
                    declared_inline_targets.add(target_id)
                source_and_connector = (
                    body[: inline_target_match.start()] + inline_target_match.group("connector")
                )
                stabilized_lines.append(f"{indent}{source_and_connector}{target_id}")
                changed = True
                continue

        match = EDGE_LABEL_IN_LINE_RE.search(body)
        if not match:
            stabilized_lines.append(line)
            continue

        source = body[: match.start()].rstrip()
        target = body[match.end() :].lstrip()
        if not source or not target:
            stabilized_lines.append(line)
            continue

        relation_id = f"Rel_{relation_index}"
        relation_index += 1
        relation_label = _clean_relation_label(match.group("label"))
        connector = match.group("connector")
        stabilized_lines.append(f'{indent}{relation_id}["{relation_label}"]:::relation')
        stabilized_lines.append(f"{indent}{source} {connector} {relation_id}")
        stabilized_lines.append(f"{indent}{relation_id} {connector} {target}")
        changed = True

    if changed and not any(re.match(r"^\s*classDef\s+relation\b", line) for line in stabilized_lines):
        stabilized_lines.append(
            "classDef relation fill:#eef2ff,stroke:#6366f1,stroke-width:1px,color:#312e81;"
        )
    return "\n".join(stabilized_lines).strip(), changed


def normalize_mermaid_code(content: str) -> tuple[str, list[str]]:
    """Normalize generated Mermaid without changing diagram semantics."""

    cleaned = strip_mermaid_code_fences(content)
    cleaned = "\n".join(line.rstrip() for line in cleaned.splitlines()).strip()
    rules: list[str] = []
    if cleaned != content.strip():
        rules.append("mermaid.trimmed_code_fences")
    if not cleaned or not _is_flowchart(cleaned):
        return cleaned, rules

    normalized_lines: list[str] = []
    changed_edge_labels = False
    for line in cleaned.splitlines():
        if line.lstrip().startswith("%%"):
            normalized_lines.append(line)
            continue
        normalized = EDGE_PIPE_LABEL_RE.sub(_quote_edge_label, line)
        if normalized != line:
            changed_edge_labels = True
        normalized_lines.append(normalized)

    normalized_code = "\n".join(normalized_lines).strip()
    normalized_code, stabilized_edge_labels = _stabilize_flowchart_edge_labels(normalized_code)
    if changed_edge_labels:
        rules.append("mermaid.quote_flowchart_edge_labels")
    if stabilized_edge_labels:
        rules.append("mermaid.stabilize_flowchart_edges")
    return normalized_code, rules

File: frontend/src/utils/mermaidSanitizer.ts
const CODE_FENCE_RE = /^```(?:mermaid)?\s*|\s*```$/gi;
const FLOWCHART_HEADER_RE = /^(?:flowchart|graph)\s+/i;
const EDGE_PIPE_LABEL_RE = /([-.=ox<>]+\|)([^|\n]+)(\|)/g;
const EDGE_LABEL_IN_LINE_RE = /([-.=ox<>]{2,})\s*\|([^|\n]+)\|\s*/;
const INLINE_EDGE_TARGET_DECL_RE = /(\s+[-.=ox<>]{2,}\s+)([A-Za-z0-9_][A-Za-z0-9_-]*(?:\[[^\n\]]+\]|\([^\n)]+\)|\{[^\n}]+\}):::[A-Za-z0-9_-]+)\s*$/;
const NODE_ID_RE = /^([A-Za-z0-9_][A-Za-z0-9_-]*)/;
const RISKY_EDGE_LABEL_RE = /[()[\]{}（）;]/;

export function stripMermaidCodeFences(content: string): string {
  const unfenced = content
    .trim()
    .replace(/^\uFEFF/, '')
    .replace(CODE_FENCE_RE, '')
    .trim();
  const smartUnwrapped = unfenced.replace(/^[“”"']+/, '').replace(/[“”"']+$/, '').trim();
  return smartUnwrapped || unfenced;
}

function isFlowchart(code: string): boolean {
  for (const line of code.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('%%')) continue;
    return FLOWCHART_HEADER_RE.test(trimmed);
  }
  return false;
}

function quoteEdgeLabel(prefix: string, labelValue: string, suffix: string): string {
  const label = labelValue.trim();
  if (label.startsWith('"') && label.endsWith('"')) {
    return `${prefix}${labelValue}${suffix}`;
  }
  if (!RISKY_EDGE_LABEL_RE.test(label)) {
    return `${prefix}${labelValue}${suffix}`;
  }
  return `${prefix}"${label.replace(/"/g, '#quot;')}"${suffix}`;
}

function cleanRelationLabel(labelValue: string): string {
  const label = labelValue.trim();
  const unquoted = label.startsWith('"') && label.endsWith('"') ? label.slice(1, -1) : label;
  return unquoted.replace(/"/g, '#quot;');
}

export function normalizeMermaidCode(content: string): string {
  const cleaned = stripMermaidCodeFences(content)
    .split('\n')
    .map((line) => line.trimEnd())
    .join('\n')
    .trim();

  if (!cleaned || !isFlowchart(cleaned)) {
    return cleaned;
  }

  return cleaned
    .split('\n')
    .map((line) => {
      if (line.trimStart().startsWith('%%')) return line;
      return line.replace(EDGE_PIPE_LABEL_RE, (_match, prefix, labelValue, suffix) =>
        quoteEdgeLabel(String(prefix), String(labelValue), String(suffix)),
      );
    })
    .join('\n')
    .trim();
}

export function stabilizeMermaidFlowchartEdgeLabels(content: string): string {
  const normalized = normalizeMermaidCode(content);
  if (!normalized || !isFlowchart(normalized)) {
    return normalized;
  }

  let relationIndex = 0;
  let changed = false;
  const declaredInlineTargets = new Set<string>();
  const lines = normalized.split('\n');
  const stabilizedLines: string[] = [];

  for (const line of lines) {
    if (line.trimStart().startsWith('%%')) {
      stabilizedLines.push(line);
      continue;
    }
    const indent = line.match(/^\s*/)?.[0] || '';
    const body = line.slice(indent.length);

    const inlineTargetMatch = body.match(INLINE_EDGE_TARGET_DECL_RE);
    if (inlineTargetMatch) {
      const targetDecl = inlineTargetMatch[2];
      const targetId = targetDecl.match(NODE_ID_RE)?.[1];
      if (targetId) {
        if (!declaredInlineTargets.has(targetId)) {
          stabilizedLines.push(`${indent}${targetDecl}`);
          declaredInlineTargets.add(targetId);
        }
        const sourceAndConnector = `${body.slice(0, inlineTargetMatch.index)}${inlineTargetMatch[1]}`;
        stabilizedLines.push(`${indent}${sourceAndConnector}${targetId}`);
        changed = true;
        continue;
      }
    }

    const match = body.match(EDGE_LABEL_IN_LINE_RE);
    if (!match) {
      stabilizedLines.push(line);
      continue;
    }

    const source = body.slice(0, match.index).trimEnd();
    const connector = match[1];
    const labelValue = match[2];
    const target = body.slice((match.index || 0) + match[0].length).trimStart();
    if (!source || !target) {
      stabilizedLines.push(line);
      continue;
    }
    const relationId = `Rel_${relationIndex++}`;
    const relationLabel = cleanRelationLabel(labelValue);
    stabilizedLines.push(`${indent}${relationId}["${relationLabel}"]:::relation`);
    stabilizedLines.push(`${indent}${source} ${connector} ${relationId}`);
    stabilizedLines.push(`${indent}${relationId} ${connector} ${target}`);
    changed = true;
  }

  if (!changed) {
    return normalized;
  }

  const hasRelationClass = stabilizedLines.some((line) => /^\s*classDef\s+relation\b/.test(line));
  if (!hasRelationClass) {
    stabilizedLines.push('classDef relation fill:#eef2ff,stroke:#6366f1,stroke-width:1px,color:#312e81;');
  }
  return stabilizedLines.join('\n').trim();
}
